plugin/trino-redshift/src/main/java/io/trino/plugin/redshift/RedshiftUnloadSplitSource.java

```diff
@@ -0,0 +1,195 @@
+
+    private boolean finished;
+
+    public RedshiftUnloadSplitSource(
+            ExecutorService executor,
+            ConnectorSession session,
+            JdbcClient jdbcClient,
+            JdbcTableHandle jdbcTableHandle,
+            List<JdbcColumnHandle> columns,
+            QueryBuilder queryBuilder,
+            RemoteQueryModifier queryModifier,
+            Optional<String> unloadLocation,
+            Optional<String> unloadAuthorization,
+            TrinoFileSystem fileSystem)
+    {
+        requireNonNull(executor, "executor is null");
+        requireNonNull(session, "session is null");
+        this.jdbcClient = requireNonNull(jdbcClient, "jdbcClient is null");
+        requireNonNull(jdbcTableHandle, "jdbcTableHandle is null");
+        requireNonNull(columns, "columns is null");
+        this.queryBuilder = requireNonNull(queryBuilder, "queryBuilder is null");
+        this.queryModifier = requireNonNull(queryModifier, "queryModifier is null");
+        this.unloadAuthorization = requireNonNull(unloadAuthorization, "unloadAuthorization is null");
+        this.fileSystem = requireNonNull(fileSystem, "fileSystem is null");
+
+        String queryFragmentId = session.getQueryId() + "-" + UUID.randomUUID();
+        this.unloadOutputPath = unloadLocation.orElseThrow() + "/" + queryFragmentId + "/";
+
+        resultSetFuture = CompletableFuture.runAsync(() -> {
+            try (Connection connection = jdbcClient.getConnection(session)) {
+                String redshiftSelectSql = buildRedshiftSelectSql(session, connection, jdbcTableHandle, columns);
+                try (PreparedStatement statement = buildUnloadSql(session, connection, columns, redshiftSelectSql, unloadOutputPath)) {
+                    // Exclusively set readOnly to false to avoid query failing with "ERROR: transaction is read-only".
+                    connection.setReadOnly(false);
+                    log.debug("Executing: %s", statement);
+                    long start = System.nanoTime();
+                    statement.execute(); // Return value of `statement.execute()` is not useful as it always return false.
+                    log.info("Redshift UNLOAD command for %s query took %s", queryFragmentId, nanosSince(start));
+                }
+            }
+            catch (SQLException e) {
+                if (e instanceof RedshiftException && e.getMessage() != null && e.getMessage().contains("The S3 bucket addressed by the query is in a different region from this cluster")) {
+                    throw new TrinoException(REDSHIFT_S3_CROSS_REGION_UNSUPPORTED, "Redshift cluster and S3 bucket in different regions is not supported", e);
+                }
+                throw new RuntimeException(e);
+            }
+        }, executor);
+    }
+
+    private String buildRedshiftSelectSql(ConnectorSession session, Connection connection, JdbcTableHandle table, List<JdbcColumnHandle> columns)
```

Comment:
move all private methods below public ones and static methods below non-static ones